include/rcpputils/time.hpp

```diff
@@ -0,0 +1,62 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCPPUTILS__TIME_HPP_
+#define RCPPUTILS__TIME_HPP_
+
+#include <chrono>
+
+#include "rcutils/time.h"
+
+#include "rcpputils/visibility_control.hpp"
+
+namespace rcpputils
+{
+
+/// Convert to rcutils duration value.
+/*
+ * \param[in] time The time to be converted to rcutils duration.
+ * \return rcutils duration value
+ * \throws std::invalid_argument if time is bigger than std::chrono::nanoseconds::max().
+ */
+template<typename DurationRepT = int64_t, typename DurationT = std::milli>
+RCPPUTILS_PUBLIC
+rcutils_duration_value_t convert_to_rcutils_duration(
+  const std::chrono::duration<DurationRepT, DurationT> & time)
+{
+  // Casting to a double representation might lose precision and allow the check below to succeed
+  // but the actual cast to nanoseconds fail. Using 1 DurationT worth of nanoseconds less than max
+  constexpr auto maximum_safe_cast_ns =
+    std::chrono::nanoseconds::max() - std::chrono::duration<DurationRepT, DurationT>(1);
+  // If period is greater than nanoseconds::max(), the duration_cast to nanoseconds will overflow
+  // a signed integer, which is undefined behavior. Checking whether any std::chrono::duration is
+  // greater than nanoseconds::max() is a difficult general problem. This is a more conservative
+  // version of Howard Hinnant's (the <chrono> guy>) response here:
+  // https://stackoverflow.com/a/44637334/2089061
+  // However, this doesn't solve the issue for all possible duration types of period.
+  // Follow-up issue: https://github.com/ros2/rclcpp/issues/1177
+  constexpr auto ns_max_as_double =
+    std::chrono::duration_cast<std::chrono::duration<double, std::chrono::nanoseconds::period>>(
+    maximum_safe_cast_ns);
+  if (time > ns_max_as_double) {
+    throw std::invalid_argument{
+            "time must be less than std::chrono::nanoseconds::max()"};
+  }
+
+  return (std::chrono::duration_cast<std::chrono::nanoseconds>(time)).count();
```

Comment:
we do not need to check `integer overflow`? i am not sure if this will never happens though.

```
  if (period_ns < std::chrono::nanoseconds::zero()) {
    throw std::runtime_error{
            "Casting timer period to nanoseconds resulted in integer overflow."};
  }
```

Reply:
```c++
  constexpr auto maximum_safe_cast_ns =
    std::chrono::nanoseconds::max() - std::chrono::duration<DurationRepT, DurationT>(1);

  constexpr auto ns_max_as_double =
    std::chrono::duration_cast<std::chrono::duration<double, std::chrono::nanoseconds::period>>(
    maximum_safe_cast_ns);
  if (time > ns_max_as_double) {
    throw std::invalid_argument{
            "time must be less than std::chrono::nanoseconds::max()"};
  }
```
Above code make sure overflow doesn't occur. It gets std::chrono::duration<double, std::chrono::nanoseconds::period> with the maximum value of int64. And compare with `time`. If overflow, exception is throwed.

Reply:
thanks, original code has this check just in case, so i was just wondering that would be needed.